### DISCUSSION TITLE
How do I get autocorrect code from offense or corrector objects (custom formatter)

### DISCUSSION THREAD

**Discussion by ghost** [rubocop/rubocop#9920] [Q&A] [answered]
I'm attempting to make a custom formatter to automatically generate code suggestions as part of a cdci process.
I'm having trouble figuring out how to coerce the values from the offense object. 

Here's what I have, (basically just a copy of the existing github formatter at the moment)

```

class GithubSuggestionFormatter < RuboCop::Formatter::GitHubActionsFormatter

  def report_offense(file, offense)
       output.printf(
      "\n::%<severity>s file=%<file>s,line=%<line>d,col=%<column>d::%<message>s\n",
      severity: github_severity(offense),
      file: file,
      line: offense.line,
      column: offense.real_column,
      message: github_escape(offense.corrector) #offense/corrector issue in question
    )
  end
end
```

**Comment by marcandre** [comment, marked as answer]
The `corrector` attribute of the offense is a `Cop::Corrector`, which itself inherits from `parser`'s [TreeRewriter](https://github.com/whitequark/parser/blob/master/lib/parser/source/tree_rewriter.rb).

You probably want to apply the corrector by calling `TreeRewriter#process` and presenting the newly obtained source code (through some come of diff?).

Alternatively you could call `TreeRewriter#as_replacements` and use that to format your suggestion.

Note: I just noticed `parser`'s official doc are [out of date](https://github.com/whitequark/parser/issues/806), so you 'll have to check the [source](https://github.com/whitequark/parser/blob/master/lib/parser/source/tree_rewriter.rb) for these methods.

**Comment by ghost** [comment]
Thank you very much. I'll have to do more experimenting to be sure it will work for my case, but I was able to get the information needed with the process method.
